File: samples/php-laravel/app/Services/OrderService.php
<?php

namespace App\Services;

use App\Repositories\OrderRepository;

class OrderService
{
    public function __construct(
        private readonly OrderRepository $orderRepository,
    ) {}

    /**
     * List all orders.
     *
     * @return array<int, array<string, mixed>>
     */
    public function listOrders(): array
    {
        return $this->orderRepository->findAll();
    }

    /**
     * Create a new order after validation.
     *
     * @param  array{product: string, quantity: int}  $data
     * @return array<string, mixed>
     */
    public function createOrder(array $data): array
    {
        if (empty($data['product'])) {
            throw new \InvalidArgumentException('Product name is required');
        }

        if (($data['quantity'] ?? 0) < 1) {
            throw new \InvalidArgumentException('Quantity must be at least 1');
        }

        return $this->orderRepository->create($data);
    }
}

File: samples/python-fastapi/app/services/order_service.py
from __future__ import annotations

from app.models import OrderCreate
from app.repositories.order_repository import OrderRepository


class OrderService:
    """Business-logic layer sitting between routers and the repository."""

    def __init__(self, repository: OrderRepository) -> None:
        self._repository = repository

    async def list_orders(self) -> list[dict]:
        """Retrieve all orders from the repository."""
        return await self._repository.find_all()

    async def create_order(self, data: OrderCreate) -> dict:
        """Validate and persist a new order."""
        if data.quantity <= 0:
            raise ValueError("quantity must be a positive integer")
        return await self._repository.create(data.model_dump())
